### PR TITLE
8323601: Improve LayoutPath.PathElement::toString

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to add an improved Improve `LayoutPath.PathElement::toString` method simplifying debugging.

Opinions and suggestions for `static PathElement sequenceElement(long start, long step)` are welcome.

### CODE REVIEW

File: test/jdk/java/foreign/TestLayoutPaths.java [hunk 1]
```diff
@@ -311,6 +311,42 @@ public void testOffsetHandle(MemoryLayout layout, PathElement[] pathElements, lo
...
+    @Test
+    public void testSequenceElementIndexToString() {
+        PathElement e = PathElement.sequenceElement(2);
+        assertEquals(e.toString(), "bound sequence element 2");
+    }
+
+    @Test
+    public void testSequenceElementRangeToString() {
+        PathElement e = PathElement.sequenceElement(2, 4);
+        assertEquals(e.toString(), "sequence range 2 + N * 4, N >= 0");
```
Comment (minborg) [inline test/jdk/java/foreign/TestLayoutPaths.java:341]: Not sure this is the best way to express the toString for the range version.

Comment (mcimadamore) [reply test/jdk/java/foreign/TestLayoutPaths.java:341]: I think there's some choices here. One way possible way to think about paths is to imagine the consequences of chaining the strings. In the proposed implementation, this would yield very good/hard to parse results, I think.

Another way would be to borrow some C syntax, to make it look more like an access expression, so that we we can see something like:

```
a.b[23].c[].d->e
```

That said, while the above looks nice to read, it's probably hard to generate, because when printing out a single element we don't know whether e.g. to emit `.` or `->` (as a separator).

But, we could still take some inspiration from that, and have the following string representations:

* `a`
* `b[23]`
* `c[]`
* `d`
* `*e`

Then if a client chains everything, they can use e.g. `/` (or `.`) as a delimiter and it will print out something sensible and compact.

Back to your question, for the sequence range expression we can use a Numpy-style syntax:

`<start index> : <step>`

So, in this case:

`[2:4]`

Which is, again, relatively compact and "composable".

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 2]
```diff
@@ -949,7 +954,8 @@ static PathElement sequenceElement() {
          */
         static PathElement dereferenceElement() {
             return new LayoutPath.PathElementImpl(PathKind.DEREF_ELEMENT,
-                    LayoutPath::derefElement);
+                    LayoutPath::derefElement,
+                    "*");
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:958]: It seems that this would result in paths like `a.b*` rather than the `*a.b`, which is correct C syntax.

Comment (minborg) [reply src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:958]: Correct. Additional logic is needed to form a correct C syntax. It would be possible to provide a method that does this.

Comment (minborg) [reply src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:958]: We could add such a method under another issue (https://bugs.openjdk.org/browse/JDK-8323746) because it will be much easier to implement it once the contemplated changes for that issue are in.

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:958]: After some more thinking, I think the toString should just mimic the expression used to create the path.
e.g. 
groupElement("foo")
or
sequenceElement()
sequenceElement(2)
sequenceElement(2, 3)
etc.

**Review by JornVernee — APPROVED** [review]
I like the latest approach. It leans on familiar names/syntax. I think that will be more intuitive than mimicking C syntax.

**Review by mcimadamore — COMMENTED** [review]
Good job
